source/neuropod/bindings/java/src/main/native/com_uber_neuropod_Neuropod.cc

```diff
@@ -215,3 +216,65 @@ JNIEXPORT jlong JNICALL Java_com_uber_neuropod_Neuropod_nativeGetGenericAllocato
     }
     return reinterpret_cast<jlong>(nullptr);
 }
+
+JNIEXPORT jobject JNICALL Java_com_uber_neuropod_Neuropod_nativeInfer(
+    JNIEnv *env, jclass, jobjectArray entryArray, jobject requestedOutputsJava, jlong modelHandle)
+{
+    try
+    {
+        // Prepare requestedOutputs
+        std::vector<std::string> requestedOutputs;
+        if (requestedOutputsJava != nullptr)
+        {
+            jsize size = env->CallIntMethod(requestedOutputsJava, java_util_ArrayList_size);
+            for (jsize i = 0; i < size; i++)
+            {
+                jstring element =
+                    static_cast<jstring>(env->CallObjectMethod(requestedOutputsJava, java_util_ArrayList_get, i));
+                requestedOutputs.emplace_back(toString(env, element));
+                env->DeleteLocalRef(element);
+            }
+        }
+
+        // Fill in NeuropodValueMap
+        jsize                      entrySize = env->GetArrayLength(entryArray);
+        neuropod::NeuropodValueMap nativeMap;
+        for (jsize i = 0; i < entrySize; i++)
+        {
+            jobject     entry = env->GetObjectArrayElement(entryArray, i);
+            std::string key =
+                toString(env, static_cast<jstring>(env->CallObjectMethod(entry, java_util_Map_Entry_getKey)));
+            jobject value        = env->CallObjectMethod(entry, java_util_Map_Entry_getValue);
+            jlong   tensorHandle = env->CallLongMethod(value, com_uber_neuropod_NeuropodTensor_getHandle);
+            if (tensorHandle == 0)
+            {
+                throw std::runtime_error("unexpected NULL tensor handle");
+            }
+            nativeMap.insert(
+                std::make_pair(key, *reinterpret_cast<std::shared_ptr<neuropod::NeuropodValue> *>(tensorHandle)));
+            env->DeleteLocalRef(entry);
+            env->DeleteLocalRef(value);
+        }
+
+        auto model       = reinterpret_cast<neuropod::Neuropod *>(modelHandle);
+        auto inferredMap = model->infer(nativeMap, requestedOutputs);
+
+        // Put data to Java Map
+        auto  ret     = env->NewObject(java_util_HashMap, java_util_HashMap_);
+        auto &entries = *inferredMap;
+        for (auto &entry : entries)
```

Comment:
You know, initially this was how you suggests. But I added extra reference to make sure that it doesn't create a copy (like auto always not reference except few cases).  I understand this is probably what user expect but I had no time to find what standard or some clarification says about this case and decided to add it explicitly.

Reply:
I think this explains it that my fear is not right https://en.cppreference.com/w/cpp/language/range-for
```
{
auto && __range = range_expression ;
for (auto __begin = begin_expr, __end = end_expr; __begin != __end; ++__begin) 
{
range_declaration = *__begin;loop_statement
}
}
```

Reply:
but at the same time it mentions some cases 
```
When used with a (non-const) object that has copy-on-write semantics, the range-based for loop may trigger a deep copy by (implicitly) calling the non-const begin() member function
that is undesirable (for instance because the loop is not actually modifying the object), it can be avoided by using std::as_const
```
not this case though